### DISCUSSION TITLE
Generalize "x is y z" support

### DISCUSSION THREAD

**Discussion by lassevk** [dotnet/csharplang#8328] [Language Ideas] [answered]
Given that the new proposal for union types extends "x is y" even further, would it be possible to get a generalized way to implement the "x is y z" pattern in our own types?

It seems the proposal for union types would lower to specific code that is only supported by union types, I think it should be considered if the proposal could be split into two, where the union type proposal would be based on and require the other one, which would propose this generalized support for "x is y z" patterns.

For instance:

    public static class StringExtensions
    {
        public static bool Is(this string s, out DateTime dt)
        {
            return DateTime.TryParse(s, out dt);
        }
    }

and then:

    string s = "2024-08-02";
    if (s is DateTime dt) Console.WriteLine(dt); // 2024-08-02 as DateTime

Obviously this would be ripe for abusing, but I think it might make patterns even more useful when used a bit more cautiously.

Musings:

* Would not override already implemented support in the language, so "if (s is object)" could not be overridden for instance (and should ideally produce a warning or error when one tries to implement such a method, this would ensure that the introduction of such a method into a class library would not make existing code behave differently if recompiled
* Would there be a separate method without the `out Type parameter` part, to support "if (x is y)", where the output variable isn't needed? I would assume that the previous method would support this and the output variable would simply be discarded, but if the act of checking the "x is y" part is relatively cheaper than "also produce z", then perhaps the extra method could be implemented and would be used in such cases
* Such implementations would make existing nuget packages, such as "OneOf", add support that make their types much more usable (but then OneOf is a union type, and the existing proposal for that might render OneOf obsolete)

*(note, I did do a search for similar discussions, but the word "is" seems to be a popular word, apologies if there is a similar post here)*

**Comment by ufcpp** [comment, marked as answer]
https://github.com/dotnet/csharplang/issues/1047
